extensions/data-transfer/portability-data-transfer-apple/src/main/java/org/datatransferproject/datatransfer/apple/photos/AppleMediaImporter.java

```diff
@@ -67,38 +84,68 @@ public ImportResult importItem(
+    final Map<String, Integer> counts = new HashMap<>();
+
+    // lower stack retry logic in data copier will not handle the skippable error case, that's why we want to build retry logic in importer itself.
+
+    // executor can either be a RetryingInMemoryIdempotentImportExecutor or an InMemoryIdempotentImportExecutor
+    // RetryingInMemoryIdempotentImportExecutor will return null for skippable error, throw for the others
+    // InMemoryIdempotentImportExecutor will throw every error (the callableImporter will throw it as well if we don't throw it here)
+
+    // if executor is a RetryingInMemoryIdempotentImportExecutor, then it will be different from the idempotentExecutor in the param, which be checked for errors in lower stack (CallableImporter),
+    // So if the error in executor is skippable, we need to clean the errors in idempotentExecutor to make sure they will not be thrown in the lower stack.
+
+    executor.executeOrThrowException(retryId, retryId, () -> {
+      importPhotosMap.clear();
+      importVideosResult.clear();
+      counts.clear();
+      idempotentExecutor.resetRecentErrors();
+
+      final int albumCount =
+              mediaInterface.importAlbums(
+                      jobId,
+                      idempotentExecutor,
+                      data.getAlbums(),
+                      DataVertical.MEDIA.getDataType());
+      importPhotosMap.putAll(
+              mediaInterface.importAllMedia(
+                      jobId,
+                      idempotentExecutor,
+                      data.getPhotos(),
+                      DataVertical.MEDIA.getDataType()));
+      importVideosResult.putAll(
+              mediaInterface.importAllMedia(
+                      jobId,
+                      idempotentExecutor,
+                      data.getVideos(),
+                      DataVertical.MEDIA.getDataType()));
+      counts.putAll(
+              new ImmutableMap.Builder<String, Integer>()
+                      .put(MediaContainerResource.ALBUMS_COUNT_DATA_NAME, albumCount)
+                      .put(
+                              MediaContainerResource.PHOTOS_COUNT_DATA_NAME,
+                              importPhotosMap.getOrDefault(ApplePhotosConstants.COUNT_KEY, 0L).intValue())
+                      .put(
+                              MediaContainerResource.VIDEOS_COUNT_DATA_NAME,
+                              importVideosResult.getOrDefault(ApplePhotosConstants.COUNT_KEY, 0L).intValue())
+                      .build());
+
+      Collection<ErrorDetail> errors = idempotentExecutor.getRecentErrors();
+      if (!errors.isEmpty() && executor instanceof RetryingInMemoryIdempotentImportExecutor) { // throw the error for retryExecutor to retry, only include the actual error message, but not the stack traces
+        throw new IOException(errors.iterator().hasNext() ? errors.iterator().next().exception().lines().findFirst().get() : ApplePhotosConstants.APPLE_PHOTOS_IMPORT_ERROR_PREFIX + " Unknown Error");
```

Comment:
Why do we not want to throw the entire stack trace? Is this logic necessary? Can we not just let the errors be thrown by the executor?

Reply:
If we let the error be thrown by idempotentExecutor midway of the import, then it will fail the whole batch even if only one file failed. 
Here we throw after the importResult was updated, and the importResult will include all the files except for the failed one.

Reply:
Oh just to make regex easier.